src/dcqc/filesystems/synapsefs.py

```diff
@@ -0,0 +1,666 @@
+        elif "already exists" in message:
+            raise DirectoryExists(message)
+        else:
+            raise  # Raise original exception as is
+
+
+class SynapseFS(FS):
+    """Construct a Synapse filesystem for
+    `PyFilesystem <https://pyfilesystem.org>`_
+
+    Args:
+        root (str): Synapse ID for a project or folder.
+        auth_token (str, optional): Synapse personal access token.
+            Defaults to None.
+        synapse_args (dict, optional): Dictionary of
+            arguments to pass to the ``Synapse()`` class.
+            Defaults to None.
+    """
+
+    NULL_BYTE = b"\x00"
+
+    SYNID_REGEX = re.compile(r"syn[0-9]+")
+
+    SUPPORTED_TYPES: dict[str, Type[Entity]]
+    SUPPORTED_TYPES = {
+        "file": File,
+        "folder": Folder,
+        "project": Project,
+    }
+
+    DEFAULT_SYNAPSE_ARGS: dict[str, Any]
+    DEFAULT_SYNAPSE_ARGS = {
+        "silent": True,
+    }
+
+    _meta = {
+        "case_insensitive": False,
+        "invalid_path_chars": "\0",
+        "network": True,
+        "read_only": False,
+        "thread_safe": True,
+        "unicode_paths": False,
+        "virtual": False,
+    }
+
+    def __init__(
+        self,
+        root: Optional[str] = None,
+        auth_token: Optional[str] = None,
+        synapse_args: Optional[dict[str, Any]] = None,
+    ) -> None:
+        super(SynapseFS, self).__init__()
+        self.auth_token = auth_token
+        self.synapse_args = synapse_args or self.DEFAULT_SYNAPSE_ARGS
+        self._local = threading.local()
+        self.root = self._resolve_root(root)
+
+    @property
+    def synapse(self) -> Synapse:
+        """Construct a thread-local Synapse client.
+
+        Returns:
+            Synapse: Authenticated Synapse client
+        """
+        if not hasattr(self._local, "synapse"):
+            # Override cache with temporary directory to avoid unwanted side effects
+            self.synapse_args["cache_root_dir"] = mkdtemp()
+            synapse = Synapse(**self.synapse_args)
+            synapse.login(authToken=self.auth_token)
+            self._local.synapse = synapse
+        # Clear the Synapse cache to ensure up-to-date
+        self._local.synapse.cache.purge(after_date=0)
+        return self._local.synapse
+
+    def is_synapse_id(self, text: str) -> bool:
+        """Check whether the given text is a Synapse ID."""
+        return self.SYNID_REGEX.fullmatch(text) is not None
+
+    def _resolve_root(self, root: Optional[str]) -> Optional[str]:
+        """Resolve the given root path (if not None) to a Synapse entity ID.
+
+        Args:
+            root (Optional[str]): _description_
+
+        Raises:
+            CreateFailed: If the root is not or does not start with a Synapse ID.
+            CreateFailed: If the root does not resolve to a project or folder.
+
+        Returns:
+            Optional[str]: A single Synapse ID to act as the root.
+        """
+        if root is None or root == "":
```

Comment:
What is the importance of allowing the file system to be rootless or not? Why might someone choose to use this one way vs. the other? Just trying to wrap my head around this part

Reply:
Great question! 

The main use case for rootless SynapseFS is the ability to open files and folders wherever they are (as long as your Synapse account has access to them). If you provide a root (_e.g._ a Synapse project), you can't open anything outside of that root. In the case of DCQC, the Synapse URIs might point anywhere, so it's preferable to not specify a root. 

Another reason is the user experience (UX) for opening files. In DCQC, all I know about a file is its URI, such as `syn://syn50555279`. If I didn't have rootless mode (which was the case initially), how would I initialize `SynapseFS` with? I eventually added support for opening a root like `syn://syn50555279/..` (_i.e._ open the given file's parent directory). However, this is clunky and unintuitive. 

Overall, I concluded that adding a rootless mode resulted in a more elegant UX: 

```python
# Without rootless mode
fs = SynapseFS("syn://syn50555279/..")
with fs.open("syn50555279") as infile:
    contents = infile.read()

# With rootless mode
fs = SynapseFS()
with fs.open("syn50555279") as infile:
    contents = infile.read()
```

Reply:
This makes a lot of sense, thanks for the response!